contracts/staking/TribalChiefSyncV2.sol

```diff
@@ -0,0 +1,191 @@
+ */
+contract TribalChiefSyncV2 {
+    ITribalChief public immutable tribalChief;
+    IAutoRewardsDistributor public immutable autoRewardsDistributor;
+    ITimelock public immutable timelock;
+
+    /// @notice a mapping from reward rates to timestamps after which they become active
+    mapping(uint256 => uint256) public rewardsSchedule;
+
+    uint256[] internal rewardsArray;
+
+    // TribalChief struct
+    struct RewardData {
+        uint128 lockLength;
+        uint128 rewardMultiplier;
+    }
+
+    constructor(
+        ITribalChief _tribalChief,
+        IAutoRewardsDistributor _autoRewardsDistributor,
+        ITimelock _timelock,
+        uint256[] memory rewards,
+        uint256[] memory timestamps
+    ) {
+        tribalChief = _tribalChief;
+        autoRewardsDistributor = _autoRewardsDistributor;
+        timelock = _timelock;
+
+        require(rewards.length == timestamps.length, "length");
+
+        uint256 lastReward = type(uint256).max;
+        uint256 lastTimestamp = block.timestamp;
+        uint256 len = rewards.length;
+        rewardsArray = new uint256[](len);
+
+        for (uint256 i = 0; i < len; i++) {
+            uint256 nextReward = rewards[i];
+            uint256 nextTimestamp = timestamps[i];
+            
+            require(nextReward < lastReward, "rewards");
+            require(nextTimestamp > lastTimestamp, "timestamp");
+            
+            rewardsSchedule[nextReward] = nextTimestamp;
+            rewardsArray[len - i - 1] = nextReward;
+
+            lastReward = nextReward;
+            lastTimestamp = nextTimestamp;
+        }
+    }
+
+    /// @notice Before an action, mass update all pools, after sync the autoRewardsDistributor
+    modifier update {
+        uint256 numPools = tribalChief.numPools();
+        uint256[] memory pids = new uint256[](numPools);
+        for (uint256 i = 0; i < numPools; i++) {
+            pids[i] = i;
+        }
+        tribalChief.massUpdatePools(pids);
+        _;
+        autoRewardsDistributor.setAutoRewardsDistribution();
+    }
+
+    /// @notice Sync a rewards rate change automatically using pre-approved map
+    function autoDecreaseRewards() external update {
+        require(isRewardDecreaseAvailable(), "time not passed");
+        uint256 tribePerBlock = nextRewardsRate();
+        tribalChief.updateBlockReward(tribePerBlock);
+        rewardsArray.pop();
+    }
+
+    function isRewardDecreaseAvailable() public view returns(bool) {
+        return nextRewardTimestamp() < block.timestamp;
+    }
+
+    function nextRewardTimestamp() public view returns(uint256) {
+        return rewardsSchedule[nextRewardsRate()];
+    }
+
+    function nextRewardsRate() public view returns(uint256) {
+        return rewardsArray[rewardsArray.length - 1];
+    }
+
+    /// @notice Sync a rewards rate change
+    function decreaseRewards(uint256 tribePerBlock, bytes32 salt) external update {
```

Comment:
prefer inheriting existing TribalChiefSync instead of copying this code for the last 4 functions.

Reply:
Plan on deprecating old contract. Rather not inherit for readabiity